lib/util.ml

```diff
@@ -69,16 +70,41 @@ module Option = struct
     | None -> default
 end
 
-module List = struct
-  let find_map f l =
-    let rec aux = function
-      | [] -> None
-      | h :: t ->
-        match f h with
-        | Some x -> Some x
-        | None -> aux t
-    in
-    aux l
+module Sexp = struct
+  type t =
+    | Atom of string
+    | List of t list
+
+  let rec equal t t' =
+    match t, t' with
+    | Atom s, Atom s' -> String.equal s s'
+    | List l, List l' -> equal_list l l'
+    | _, _ -> false
+  and equal_list l l' =
+    match l, l' with
+    | [], [] -> true
+    | hd::tl, hd'::tl' -> equal hd hd' && equal_list tl tl'
+    | _, _ -> false
+
+  module Canonical = struct
+    let to_buffer ~buf sexp =
+      let rec loop = function
+        | Atom str ->
+          Buffer.add_string buf (string_of_int (String.length str));
+          Buffer.add_string buf ":";
+          Buffer.add_string buf str
+        | List e ->
+          Buffer.add_char buf '(';
+          ignore (List.map loop e);
+          Buffer.add_char buf ')'
+      in
+      ignore (loop sexp)
```

Comment:
This is not the sexp format.
Lists are separated by whitespaces, there is no length prefix and use quotes and backslashes to escape `"abc\"def"`.